lib/schema-controller.js

```diff
@@ -105,7 +105,10 @@ class SchemaController {
    * @param {object} serverOptions the fastify server options
    */
   setupValidator (serverOptions) {
-    const isReady = this.validatorCompiler !== undefined && !this.addedSchemas
+    // WARNING: isCustomValidatorCompiler will be true if there is a custom builder OR if there is a custom compiler
+    // installed directly via setValidatorCompiler, leading to this messy condition.  Compilers installed via
+    // setValidatorCompiler will not be invalidated and recreated if schemas are added.
+    const isReady = this.validatorCompiler !== undefined && (this.isCustomValidatorCompiler || !this.addedSchemas)
```

Comment:
> For your discussion: as a new user of the framework, I would have much preferred if `setSchemaValidator` didn't exist or was marked deprecated and bad, because it is unsafe and trying to use it wasted a bunch of time. If I had found the other mechanism was there (schema controller opts) then I would have just used that. Unfortunately sample code and google hits made me find the old ("easy" but also dangerous) mechanism and I didn't know it.

@Eomm is there any plan to drop this API in v5?

So far I believe it should be ok to do this; we just need to verify the risk is documented somewhere.

Reply:
@Eomm this is resolved from my point of view since the "legacy" API no longer breaks after this fix, so I don't care if you deprecate it or not :) but it seems more like something for you to track. 

i.e. I think this is for you to resolve.